pkg/security/auth.go

```diff
@@ -205,42 +205,77 @@ func getCertificatePrincipals(cert *x509.Certificate) []string {
 func GetCertificateUserScope(
 	peerCert *x509.Certificate,
 ) (userScopes []CertificateUserScope, _ error) {
+	collectFn := func(userScope CertificateUserScope) (halt bool, err error) {
+		userScopes = append(userScopes, userScope)
+		return false, nil
+	}
+	err := forEachCertificateUserScope(peerCert, collectFn)
+	return userScopes, err
+}
+
+// CertificateUserScopeContainsFunc returns true if the given function returns
+// true for any of the scopes in the client certificate.
+func CertificateUserScopeContainsFunc(
+	peerCert *x509.Certificate, fn func(CertificateUserScope) bool,
+) (ok bool, _ error) {
+	res := false
```

Comment:
this escapes to the heap (?) but could be sync.Pooled in principle.

Reply:
This doesn't escape.